File: ui/src/transaction-csv.ts
import { command } from "./api";
import {
  digest,
  type LedgerScope,
  type Matching,
} from "./transaction-ledger-types";

export type CsvPolicy = "reject" | "allow_selected";
export type CsvRequest = {
  selection: Omit<LedgerScope, "page_size">;
  non_accepted: CsvPolicy;
};
type CsvColumn = {
  name: string;
  logical_type: string;
  prefix: string;
  nullable: boolean;
  meaning: string;
};
export type CsvDictionary = {
  schema_version: 1;
  encoding: string;
  delimiter: string;
  record_terminator: string;
  quote_policy: string;
  null_literal: string;
  columns: CsvColumn[];
  limitations: string[];
};
export type CsvIdentity = {
  workspace_revision: number;
  request: CsvRequest;
  matching: Matching;
  selection_sha256: string;
  format: "typed_literal_v1";
  format_sha256: string;
  dictionary: CsvDictionary;
  row_count: number;
  bytes: number;
  sha256: string;
};
export type CsvExport = CsvIdentity & { schema_version: 1; csv: string };
export const CSV_FORMAT = "typed_literal_v1" as const;
// Exact v1 Rust dictionary identity. Changing its meanings or column order needs
// an explicitly supported contract, not a silently accepted response dictionary.
const FORMAT_SHA256 =
  "a79b97802d789b1aa0e6828823aa18f5826bc03ac7613369f8ec0d943bcb73c4";
const MAX_BYTES = 256 * 1024 * 1024;
export function csvRequest(scope: LedgerScope, policy: CsvPolicy): CsvRequest {
  return {
    selection: { query: scope.query, filter: scope.filter, order: scope.order },
    non_accepted: policy,
  };
}
export const exactKeys = (value: unknown, names: string[]): boolean =>
  !!value &&
  typeof value === "object" &&
  !Array.isArray(value) &&
  Object.keys(value).sort().join(",") === [...names].sort().join(",");
export function equalContract(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (Array.isArray(a) && Array.isArray(b))
    return a.length === b.length && a.every((v, i) => equalContract(v, b[i]));
  if (!a || !b || typeof a !== "object" || typeof b !== "object") return false;
  const aa = a as Record<string, unknown>,
    bb = b as Record<string, unknown>;
  return (
    Object.keys(aa).length === Object.keys(bb).length &&
    Object.keys(aa).every(
      (k) => Object.hasOwn(bb, k) && equalContract(aa[k], bb[k]),
    )
  );
}
const unsigned = (value: unknown): value is number =>
  Number.isSafeInteger(value) && Number(value) >= 0;
const boundedText = (value: unknown): value is string =>
  typeof value === "string" && value.length <= 1024;
export async function sha256(bytes: Uint8Array<ArrayBuffer>): Promise<string> {
  if (!globalThis.crypto?.subtle)
    throw new Error(
      "WebCrypto is unavailable; export integrity could not be verified.",
    );
  return Array.from(
    new Uint8Array(await crypto.subtle.digest("SHA-256", bytes)),
    (n) => n.toString(16).padStart(2, "0"),
  ).join("");
}
function sortedJson(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortedJson);
  if (value && typeof value === "object")
    return Object.fromEntries(
      Object.entries(value)
        .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
        .map(([k, v]) => [k, sortedJson(v)]),
    );
  return value;
}
export async function validateCsvIdentity(
  value: CsvIdentity,
  request: CsvRequest,
  revision: number,
  count: number,
  matching: Matching,
) {
  const d = value?.dictionary;
  if (
    !value ||
    value.workspace_revision !== revision ||
    value.row_count !== count ||
    !unsigned(value.row_count) ||
    !equalContract(value.request, request) ||
    !equalContract(value.matching, matching) ||
    value.format !== CSV_FORMAT ||
    value.format_sha256 !== FORMAT_SHA256 ||
    !digest(value.selection_sha256) ||
    !digest(value.sha256) ||
    !unsigned(value.bytes) ||
    value.bytes > MAX_BYTES ||
    !exactKeys(d, [
      "schema_version",
      "encoding",
      "delimiter",
      "record_terminator",
      "quote_policy",
      "null_literal",
      "columns",
      "limitations",
    ]) ||
    d.schema_version !== 1 ||
    ![
      d.encoding,
      d.delimiter,
      d.record_terminator,
      d.quote_policy,
      d.null_literal,
    ].every(boundedText) ||
    !Array.isArray(d.columns) ||
    d.columns.length !== 15 ||
    d.columns.some(
      (c) =>
        !exactKeys(c, [
          "name",
          "logical_type",
          "prefix",
          "nullable",
          "meaning",
        ]) ||
        typeof c.nullable !== "boolean" ||
        ![c.name, c.logical_type, c.prefix, c.meaning].every(boundedText),
    ) ||
    !Array.isArray(d.limitations) ||
    d.limitations.length !== 5 ||
    !d.limitations.every(boundedText)
  )
    throw new Error(
      "CSV export identity, format, dictionary, scope or count did not match.",
    );
  const formatHash = await sha256(
    new TextEncoder().encode(JSON.stringify(sortedJson([value.format, d]))),
  );
  if (formatHash !== FORMAT_SHA256)
    throw new Error("CSV data dictionary did not match the supported format.");
}

/** Validate the fixed quoted transport format without numeric/date conversion or
 * materializing a second ledger. Rust remains the only transaction serializer. */
function validateCsvRows(
  csv: string,
  dictionary: CsvDictionary,
  expected: number,
) {
  if (!csv.startsWith("\ufeff"))
    throw new Error("CSV export is missing its UTF-8 marker.");
  let at = 1,
    rows = 0;
  while (at < csv.length) {
    for (let column = 0; column < dictionary.columns.length; column++) {
      if (csv[at++] !== '"')
        throw new Error("CSV export has an invalid quoted field.");
      const start = at;
      let closed = false;
      while (at < csv.length) {
        if (csv[at++] !== '"') continue;
        if (csv[at] === '"') {
          at++;
          continue;
        }
        closed = true;
        break;
      }
      if (!closed) throw new Error("CSV export is truncated.");
      const end = at - 1,
        definition = dictionary.columns[column];
      if (rows === 0) {
        if (csv.slice(start, end) !== definition.name)
          throw new Error("CSV export header does not match its dictionary.");
      } else {
        const missing =
          definition.nullable &&
          end - start === 4 &&
          csv.startsWith("null", start);
        if (!missing && !csv.startsWith(definition.prefix, start))
          throw new Error(
            "CSV export field is missing its declared literal prefix.",
          );
      }
      if (column < dictionary.columns.length - 1) {
        if (csv[at++] !== ",")
          throw new Error("CSV export column count did not match.");
      } else if (csv.slice(at, at + 2) !== "\r\n")
        throw new Error("CSV export record terminator did not match.");
      else at += 2;
    }
    rows++;
    if (rows > expected + 1)
      throw new Error("CSV export contains more rows than declared.");
  }
  if (rows !== expected + 1)
    throw new Error("CSV export row count did not match its content.");
}
export async function readTransactionCsv(
  scope: LedgerScope,
  policy: CsvPolicy,
  revision: number,
  count: number,
  matching: Matching,
) {
  const request = csvRequest(scope, policy);
  const value = await command<CsvExport>({
    action: "export_transaction_csv",
    request,
    expected_revision: revision,
  });
  if (
    !exactKeys(value, [
      "schema_version",
      "workspace_revision",
      "request",
      "matching",
      "selection_sha256",
      "format",
      "format_sha256",
      "dictionary",
      "row_count",
      "bytes",
      "sha256",
      "csv",
    ]) ||
    value.schema_version !== 1 ||
    typeof value.csv !== "string" ||
    value.csv.length > MAX_BYTES
  )
    throw new Error("CSV export response is unavailable or unsupported.");
  await validateCsvIdentity(value, request, revision, count, matching);
  const bytes = new TextEncoder().encode(value.csv);
  if (bytes.length !== value.bytes || (await sha256(bytes)) !== value.sha256)
    throw new Error("CSV export bytes did not match their identity.");
  validateCsvRows(value.csv, value.dictionary, count);
  return value;
}

File: crates/core/src/transaction_csv.rs
//! Explicit reversible CSV presentation. Raw typed JSON remains a separate export.
use crate::{
    domain::{ReviewState, Transaction},
    literal_search::LiteralMatching,
    require,
    transaction_export::TransactionExportRequest,
    Error, Result,
};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::io::{self, Write};

pub const MAX_CSV_BYTES: usize = 256 * 1024 * 1024;
const BOM: &[u8] = b"\xef\xbb\xbf";
const HEADERS: [&str; 15] = [
    "workspace_revision",
    "id",
    "version",
    "account",
    "date",
    "posting_date",
    "description",
    "amount",
    "currency",
    "balance",
    "anchor",
    "review",
    "duplicate_candidates",
    "transfer_peer",
    "merchant",
];

#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum NonAcceptedCsvPolicy {
    Reject,
    AllowSelected,
}
#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TransactionCsvFormat {
    TypedLiteralV1,
}
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct TransactionCsvRequest {
    pub selection: TransactionExportRequest,
    pub non_accepted: NonAcceptedCsvPolicy,
}
#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CsvLogicalType {
    Text,
    ExactDecimalText,
    CalendarDateText,
    UnsignedInteger,
    SourceAnchorJson,
    StringArrayJson,
}
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct TransactionCsvColumn {
    pub name: String,
    pub logical_type: CsvLogicalType,
    pub prefix: String,
    pub nullable: bool,
    pub meaning: String,
}
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct TransactionCsvDictionary {
    pub schema_version: u32,
    pub encoding: String,
    pub delimiter: String,
    pub record_terminator: String,
    pub quote_policy: String,
    pub null_literal: String,
    pub columns: Vec<TransactionCsvColumn>,
    pub limitations: Vec<String>,
}
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct TransactionCsvExport {
    pub schema_version: u32,
    pub workspace_revision: u64,
    pub request: TransactionCsvRequest,
    pub matching: LiteralMatching,
    /// Same selection identity as complete raw JSON, independent of CSV presentation.
    pub selection_sha256: String,
    pub format: TransactionCsvFormat,
    /// Versioned format and full dictionary, independent of selected row content.
    pub format_sha256: String,
    pub dictionary: TransactionCsvDictionary,
    pub row_count: u64,
    pub bytes: u64,
    /// Exact CSV bytes including UTF-8 BOM, doubled quotes and CRLF separators.
    pub sha256: String,
    pub csv: String,
}

pub(crate) fn dictionary() -> TransactionCsvDictionary {
    use CsvLogicalType::*;
    let definitions = [
        (
            UnsignedInteger,
            "uint:",
            false,
            "Workspace revision captured in the same canonical read snapshot.",
        ),
        (
            Text,
            "text:",
            false,
            "Stable canonical transaction identifier; never an account number.",
        ),
        (
            UnsignedInteger,
            "uint:",
            false,
            "Exact canonical transaction version.",
        ),
        (
            Text,
            "text:",
            false,
            "Original account text, including leading zeros.",
        ),
        (
            CalendarDateText,
            "date:",
            false,
            "Transaction calendar date YYYY-MM-DD; no time zone is implied.",
        ),
        (
            CalendarDateText,
            "date:",
            true,
            "Posting calendar date YYYY-MM-DD when available; no time zone is implied.",
        ),
        (
            Text,
            "text:",
            false,
            "Original transaction description; formulas remain literal text behind the visible prefix.",
        ),
        (
            ExactDecimalText,
            "decimal:",
            false,
            "Exact signed canonical amount string with up to eight decimal places; not a floating-point cell.",
        ),
        (
            Text,
            "text:",
            false,
            "Canonical three-letter currency; no currency conversion.",
        ),
        (
            ExactDecimalText,
            "decimal:",
            true,
            "Exact available balance string; null is distinct from zero.",
        ),
        (
            SourceAnchorJson,
            "json:",
            false,
            "Complete canonical SourceAnchor JSON; provenance reference, not an acceptance claim.",
        ),
        (
            Text,
            "text:",
            false,
            "Canonical review state: pending, accepted, rejected or deferred.",
        ),
        (
            StringArrayJson,
            "json:",
            false,
            "Exact duplicate candidate ID array; no repeated transactions are removed.",
        ),
        (
            Text,
            "text:",
            true,
            "Canonical transfer peer ID when present; export does not change or exclude transfers.",
        ),
        (
            Text,
            "text:",
            true,
            "Canonical merchant value when present; null and empty text remain distinct.",
        ),
    ];
    let columns = HEADERS
        .iter()
        .zip(definitions)
        .map(
            |(name, (logical_type, prefix, nullable, meaning))| TransactionCsvColumn {
                name: (*name).into(),
                logical_type,
                prefix: prefix.into(),
                nullable,
                meaning: meaning.into(),
            },
        )
        .collect();
    TransactionCsvDictionary {
        schema_version: 1,
        encoding: "utf-8-bom".into(),
        delimiter: ",".into(),
        record_terminator: "CRLF".into(),
        quote_policy: "all_fields_double_quoted_with_doubled_internal_quotes".into(),
        null_literal: "null".into(),
        columns,
        limitations:vec![
            "Visible alphabetic type prefixes make this a literal presentation; amounts and dates are not spreadsheet numeric/date cells.".into(),
            "Remove only the declared prefix when decoding. The exact unprefixed token null means missing; text:null and text: mean present text.".into(),
            "Raw typed JSON remains the machine-readable canonical export. CSV has no native type system; downstream edits, prefix removal and re-import settings can change interpretation.".into(),
            "No universal spreadsheet safety or tested Excel/LibreOffice behavior is claimed. Do not strip prefixes and then open untrusted values as formulas.".into(),
            "Literal text cells reject controls except tab, CR and LF. JSON cells preserve JSON escapes; any remaining unsupported literal control is also refused.".into(),
        ],
    }
}

fn cell(prefix: &str, raw: &str, limit: usize) -> Result<String> {
    require(
        prefix
            .len()
            .checked_add(raw.len())
            .is_some_and(|n| n <= limit),
        "CSV cell exceeds byte limit",
    )?;
    require(
        !raw.chars()
            .any(|c| c.is_control() && !matches!(c, '\t' | '\r' | '\n')),
        "CSV literal text contains an unsupported control character; use raw typed JSON",
    )?;
    let mut value = String::new();
    value
        .try_reserve_exact(prefix.len() + raw.len())
        .map_err(|_| Error::Validation("CSV cell allocation is unavailable".into()))?;
    value.push_str(prefix);
    value.push_str(raw);
    Ok(value)
}
fn optional_cell(prefix: &str, raw: Option<&str>, limit: usize) -> Result<String> {
    raw.map_or_else(|| Ok("null".into()), |v| cell(prefix, v, limit))
}
struct LimitedBytes {
    bytes: Vec<u8>,
    limit: usize,
}
impl Write for LimitedBytes {
    fn write(&mut self, input: &[u8]) -> io::Result<usize> {
        self.bytes
            .len()
            .checked_add(input.len())
            .filter(|n| *n <= self.limit)
            .ok_or_else(|| io::Error::other("CSV export byte limit exceeded; narrow the scope"))?;
        self.bytes.extend_from_slice(input);
        Ok(input.len())
    }
    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}
fn json_cell(value: &impl Serialize, limit: usize) -> Result<String> {
    let mut writer = LimitedBytes {
        bytes: Vec::new(),
        limit: limit.saturating_sub(5),
    };
    serde_json::to_writer(&mut writer, value)?;
    let value = String::from_utf8(writer.bytes)
        .map_err(|_| Error::Validation("CSV JSON cell is not UTF-8".into()))?;
    cell("json:", &value, limit)
}
/// Visit one owned cell at a time; never materialize a second complete row table.
fn visit_cells(
    row: &Transaction,
    revision: u64,
    limit: usize,
    mut write: impl FnMut(&str) -> Result<()>,
) -> Result<()> {
    write(&cell("uint:", &revision.to_string(), limit)?)?;
    write(&cell("text:", &row.id, limit)?)?;
    write(&cell("uint:", &row.version.to_string(), limit)?)?;
    write(&cell("text:", &row.account, limit)?)?;
    write(&cell("date:", &row.date, limit)?)?;
    write(&optional_cell("date:", row.posting_date.as_deref(), limit)?)?;
    write(&cell("text:", &row.description, limit)?)?;
    write(&cell("decimal:", &row.amount, limit)?)?;
    write(&cell("text:", &row.currency, limit)?)?;
    write(&optional_cell("decimal:", row.balance.as_deref(), limit)?)?;
    write(&json_cell(&row.anchor, limit)?)?;
    let review = match row.review {
        ReviewState::Pending => "pending",
        ReviewState::Accepted => "accepted",
        ReviewState::Rejected => "rejected",
        ReviewState::Deferred => "deferred",
    };
    write(&cell("text:", review, limit)?)?;
    write(&json_cell(&row.duplicate_candidates, limit)?)?;
    write(&optional_cell(
        "text:",
        row.transfer_peer.as_deref(),
        limit,
    )?)?;
    write(&optional_cell("text:", row.merchant.as_deref(), limit)?)
}
fn add_bytes(total: &mut usize, count: usize, limit: usize) -> Result<()> {
    *total = total
        .checked_add(count)
        .filter(|n| *n <= limit)
        .ok_or_else(|| {
            Error::Validation("Complete CSV export exceeds byte limit; narrow the scope".into())
        })?;
    Ok(())
}
fn field_bytes(raw: &str) -> Result<usize> {
    raw.len()
        .checked_add(raw.bytes().filter(|b| *b == b'"').count())
        .and_then(|n| n.checked_add(2))
        .ok_or_else(|| Error::Validation("CSV field size overflow".into()))
}
pub(crate) fn encode(
    rows: &[Transaction],
    revision: u64,
    policy: NonAcceptedCsvPolicy,
    limit: usize,
) -> Result<String> {
    require(
        policy == NonAcceptedCsvPolicy::AllowSelected
            || rows.iter().all(|row| row.review == ReviewState::Accepted),
        "CSV selection contains non-accepted records; explicitly allow the selected review states or narrow the scope",
    )?;
    // Exact preflight counts the chosen quote policy before allocating the final artifact.
    let mut expected = 0;
    add_bytes(&mut expected, BOM.len(), limit)?;
    for header in HEADERS {
        add_bytes(&mut expected, field_bytes(header)?, limit)?;
    }
    add_bytes(&mut expected, HEADERS.len() - 1 + 2, limit)?;
    for row in rows {
        visit_cells(row, revision, limit, |value| {
            add_bytes(&mut expected, field_bytes(value)?, limit)
        })?;
        add_bytes(&mut expected, HEADERS.len() - 1 + 2, limit)?;
    }
    let mut bytes = Vec::new();
    bytes
        .try_reserve_exact(expected)
        .map_err(|_| Error::Validation("CSV artifact allocation is unavailable".into()))?;
    let mut sink = LimitedBytes { bytes, limit };
    sink.write_all(BOM)?;
    let mut writer = csv::WriterBuilder::new()
        .quote_style(csv::QuoteStyle::Always)
        .terminator(csv::Terminator::CRLF)
        .double_quote(true)
        .from_writer(sink);
    writer.write_record(HEADERS)?;
    for row in rows {
        visit_cells(row, revision, limit, |value| {
            writer.write_field(value)?;
            Ok(())
        })?;
        writer.write_record(std::iter::empty::<&str>())?;
    }
    writer.flush()?;
    let sink = writer.into_inner().map_err(|e| Error::Io(e.into_error()))?;
    require(
        sink.bytes.len() == expected,
        "CSV serializer differs from byte preflight",
    )?;
    String::from_utf8(sink.bytes).map_err(|_| Error::Validation("CSV export is not UTF-8".into()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bounded_sink_refuses_oversized_write_without_retaining_a_partial_input() {
        let mut sink = LimitedBytes {
            bytes: Vec::new(),
            limit: 8,
        };
        sink.write_all(b"abc").unwrap();
        assert!(sink.write_all(b"123456").is_err());
        assert_eq!(sink.bytes, b"abc");
        sink.write_all(b"12345").unwrap();
        assert_eq!(sink.bytes, b"abc12345");
        assert!(sink.write_all(b"x").is_err());
        assert_eq!(sink.bytes.len(), 8);
    }
}
